fusion.md

```diff
@@ -0,0 +1,1396 @@
+
+
+Some Documentation for Filter Steps
+---
+
+### FilterStep objects
+
+
+
+A FilterStep is the object that GCAMFusion uses to search a single GCAM
+container's data vector.  It can optionally specify a data name to match which
+is compared against the `Data::mDataName`.  If the data name in the FilterStep
+is empty it is assumed to match *any* data name.  
+
+The other optional parameter is a [Filter](#filter-objects).  Filter objects are
+valid for search targets that are containers for other objects.  Such containers
+are indicated with the `ARRAY` or `CONTAINER` flag.  If specified, a filter can
+be used to select any single element of the matched container.  If no filter is
+specified, it is assumed to be `NoFilter`, which selects the entire
+container.  Note that if a Filter other than `NoFilter` is set, and the matched
+object is not a container (i.e., has the `SIMPLE` flag, then the match will be
+rejected even if the data name matches.  
+
+In addition if no data name and no filter is set then not only does this
+FilterStep match all data but it also enables special "descendant step"
+traversal behavior in GCAMFusion where the next FilterStep can be matched zero
+or more containers down.  This is analogous to the `//` operator in XPath
+queries.  For example, if a 'sector' object has 'subsector' children, which in
+turn have 'technology' children of their own, then `sector//share-weight` will
+find data named 'share-weight' at both the subector and technology levels.  It
+would also find a share-weight object contained in the sector itself if there
+were any, but in this example there are no such matches; share weights are only
+defined for subsectors and technologies.
+<!-- XXX: Check if this last sentence is accurate: would sector//share-weight -->
+<!-- match a share weight that is a direct child of the sector (if there were -->
+<!-- one)? -->
+
+
+
+### Filter objects
+
+A Filter object allows GCAMFusion to select a subset of a data object that is a
+container for other objects. The available Filters are:
+
+* NoFilter: matches all elements
+* NamedFilter: Calls the -&gt;getName() method on the container to compare in it's predicate. `SIMPLE` and `ARRAY` will never match.
+* YearFilter: Calls the -&gt;getYear() method if the data is `CONTAINER` or uses the `Modeltime` to convert period index to year if the data is `ARRAY` and compares that year in it's predicate. `SIMPLE` data will never match.
+* IndexFilter: Computes the index offset of each element in an `ARRAY` or `CONTAINER` Data for comparison in it's predicate. `SIMPLE` Data will never match.
+
+
+
+### Predicates
+
+
+
+A predicate is a way to test whether a year, or name, or index, etc matches a
+value the user was looking for. Currently predicates can only operate on string
+and int. If a predicate that is doing a string comparison is given an int to
+match (i.e. called from a YearFilter) it will always return false, and vice
+versa. The available predicates are:
```

Comment:
We say that a string comparison predicate will always fail to match an int.  Is the reverse true?  That is, will an int comparison predicate always fail to match a string (as I've written here)?

Reply:
Yup